internal/local/structure.go

```diff
@@ -0,0 +1,222 @@
+	"encoding/json"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+// GraphQLStructure represents the Realm app graphql structure
+type GraphQLStructure struct {
+	Config          map[string]interface{}   `json:"config,omitempty"`
+	CustomResolvers []map[string]interface{} `json:"custom_resolvers,omitempty"`
+}
+
+type SecretsStructure struct {
+	AuthProviders map[string]map[string]string `json:"auth_providers,omitempty"`
+	Services      map[string]map[string]string `json:"services,omitempty"`
+}
+
+// ServiceStructure represents the Realm app service structure
+type ServiceStructure struct {
+	Config           map[string]interface{}   `json:"config,omitempty"`
+	IncomingWebhooks []map[string]interface{} `json:"incoming_webhooks"`
+	Rules            []map[string]interface{} `json:"rules"`
+}
+
+func parseEnvironments(rootDir string) (map[string]map[string]interface{}, error) {
+	out := map[string]map[string]interface{}{}
+
+	dw := directoryWalker{
+		path:      filepath.Join(rootDir, NameEnvironments),
+		onlyFiles: true,
+	}
+	if err := dw.walk(func(file os.FileInfo, path string) error {
+		o, err := parseJSON(path)
+		if err != nil {
+			return err
+		}
+		out[file.Name()] = o
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+
+	if len(out) == 0 {
+		return nil, nil
+	}
+	return out, nil
+}
+
+func parseFunctions(rootDir string) ([]map[string]interface{}, error) {
+	if ok, err := fileExists(rootDir); err != nil {
+		return nil, err
+	} else if !ok {
+		return nil, nil
+	}
+
+	var out []map[string]interface{}
+
+	dw := directoryWalker{path: rootDir, onlyDirs: true}
+	if walkErr := dw.walk(func(file os.FileInfo, path string) error {
+		if strings.Contains(path, nameNodeModules) {
+			return nil // skip node_modules
+		}
+
+		config, configErr := parseJSONConfig(path)
+		if configErr != nil {
+			return configErr
+		}
+
+		src, srcErr := parseJavascriptSource(path)
+		if srcErr != nil {
+			return srcErr
+		}
+
+		out = append(out, map[string]interface{}{
+			NameConfig: config,
+			NameSource: src,
+		})
+		return nil
+	}); walkErr != nil {
+		return nil, walkErr
+	}
+
+	return out, nil
+}
+
+func parseGraphQL(rootDir string) (GraphQLStructure, bool, error) {
+	dir := filepath.Join(rootDir, NameGraphQL)
+
+	if ok, err := fileExists(dir); err != nil {
+		return GraphQLStructure{}, false, err
+	} else if !ok {
+		return GraphQLStructure{}, false, nil
+	}
+
+	config, configErr := parseJSONConfig(dir)
+	if configErr != nil {
+		return GraphQLStructure{}, false, configErr
+	}
+
+	customResolvers, customResolversErr := parseJSONFiles(filepath.Join(dir, NameCustomResolvers))
+	if customResolversErr != nil {
+		return GraphQLStructure{}, false, customResolversErr
+	}
+
+	return GraphQLStructure{config, customResolvers}, true, nil
+}
+
+func parseJavascript(rootDir string, file File) (string, error) {
+	src, err := ioutil.ReadFile(filepath.Join(rootDir, file.String()))
+	if err != nil {
+		return "", err
+	}
+	return string(src), nil
```

Comment:
more of a question - do we actually need this as a string? or can we keep this as a []byte?

Reply:
we do actually need this as a string in our current server implementation, i was burned by this a few weeks back in a nasty test